core/src/main/java/org/apache/iceberg/rest/requests/UpdateTableRequest.java

```diff
@@ -0,0 +1,399 @@
+    private final List<MetadataUpdate> updates = Lists.newArrayList();
+    private final Set<String> changedRefs = Sets.newHashSet();
+    private final boolean isReplace;
+    private boolean addedSchema = false;
+    private boolean setSchemaId = false;
+    private boolean addedSpec = false;
+    private boolean setSpecId = false;
+    private boolean setOrderId = false;
+
+    public Builder(TableMetadata base, boolean isReplace) {
+      this.base = base;
+      this.isReplace = isReplace;
+    }
+
+    private Builder require(UpdateRequirement requirement) {
+      Preconditions.checkArgument(requirement != null, "Invalid requirement: null");
+      requirements.add(requirement);
+      return this;
+    }
+
+    private Builder requireCreate() {
+      return require(new UpdateRequirement.AssertTableDoesNotExist());
+    }
+
+    private Builder requireTableUUID(String uuid) {
+      Preconditions.checkArgument(uuid != null, "Invalid required UUID: null");
+      return require(new UpdateRequirement.AssertTableUUID(uuid));
+    }
+
+    private Builder requireRefSnapshotId(String ref, Long snapshotId) {
+      return require(new UpdateRequirement.AssertRefSnapshotID(ref, snapshotId));
+    }
+
+    private Builder requireLastAssignedFieldId(int fieldId) {
+      return require(new UpdateRequirement.AssertLastAssignedFieldId(fieldId));
+    }
+
+    private Builder requireCurrentSchemaId(int schemaId) {
+      return require(new UpdateRequirement.AssertCurrentSchemaID(schemaId));
+    }
+
+    private Builder requireLastAssignedPartitionId(int partitionId) {
+      return require(new UpdateRequirement.AssertLastAssignedPartitionId(partitionId));
+    }
+
+    private Builder requireDefaultSpecId(int specId) {
+      return require(new UpdateRequirement.AssertDefaultSpecID(specId));
+    }
+
+    private Builder requireDefaultSortOrderId(int orderId) {
+      return require(new UpdateRequirement.AssertDefaultSortOrderID(orderId));
+    }
+
+    public Builder update(MetadataUpdate update) {
+      Preconditions.checkArgument(update != null, "Invalid update: null");
+      updates.add(update);
+
+      // add requirements based on the change
+      if (update instanceof MetadataUpdate.SetSnapshotRef) {
+        update((MetadataUpdate.SetSnapshotRef) update);
+      } else if (update instanceof MetadataUpdate.AddSchema) {
+        update((MetadataUpdate.AddSchema) update);
+      } else if (update instanceof MetadataUpdate.SetCurrentSchema) {
+        update((MetadataUpdate.SetCurrentSchema) update);
+      } else if (update instanceof MetadataUpdate.AddPartitionSpec) {
+        update((MetadataUpdate.AddPartitionSpec) update);
+      } else if (update instanceof MetadataUpdate.SetDefaultPartitionSpec) {
+        update((MetadataUpdate.SetDefaultPartitionSpec) update);
+      } else if (update instanceof MetadataUpdate.SetDefaultSortOrder) {
+        update((MetadataUpdate.SetDefaultSortOrder) update);
+      }
```

Comment:
you might want an `else` to throw at the end otherwise future updates (if there are any) may fall through.

Reply:
Other updates are okay. These methods just add requirements, so things like `AddSnapshot` are supposed to fall through without changing anything.